proposals/NNNN-constructors.md

```diff
@@ -0,0 +1,126 @@
+# Constructors
+
+* Proposal: [NNNN](NNNN-constructors.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Chris Bieneman](https://github.com/llvm-beanz)
+* Status: **Under Consideration**
+* Planned Version: 202y
+
+## Introduction
+
+As user defined data structures get more common and complex, the absence of
+constructors and destructors is more and more awkward. This proposal seeks to
+bring the full features of C++ initialization and destruction into HLSL.
+
+## Motivation
+
+Constructors are incredibly useful for ensuring data initialization and
+implementing common patterns like Resource-Acquisition-Is-Initialization (RAII).
+
+## Proposed solution
+
+Introduce the following features to HLSL:
+* C++ Constructors and Destructors
+* C++ 11 Delegating Constructors
+* C++ 11 Default and Deleted methods
+* C++ 11 Default initializers
+* Clang __attribute__((constructor)) and __attribute__((destructor))
+
+Constructor syntax will match C++ 11 constructor syntax including support for
+delegating constructors. It will be available on all user-defined data types
+that support constructors in C++.
+
+## Detailed design
+
+### Global Construction
+
+HLSL global constructors will be ordered following C++ conventions.
+Constant-evaluated initializers occur first, followed by non-constant
+initializers which are ordered by declaration order in the translation unit.
+
+Each entry function will call the global constructors before executing any
+further instructions and the global destructor immediately before returning.
+
+In library shaders, the entry translation unit's global constructors are
+executed after the linked library global constructors, and the entry's global
+destructors are executed before linked library global destructors.
+
+Non-trivial global constructors **will not** be stripped from library shaders.
+Meaning, an unused global with a non-trivial constructor will be linked into the
+final linked shader binary. This conforms to the C++ standard definitions for

```

Comment:
Other than breaking with the ISO C++ standard, is there a reason we would not strip unused global constructors? Are we opening ourselves up to customers writing a post-link optimizer to getting rid of these constructors?

Reply:
If you want a global registration to occur on every shader run always, the simple approach would be a global constructor that doesn't get stripped.

If instead you have something that you only want to happen when it is used, the standard C++ approach is a singleton pattern.

Having the compiler not strip unused constructors enables the global-registration use case, but doesn't prevent a user from writing code that depends on use for initialization, it is just a different pattern.